### PR TITLE
[Data] Support HeteroGraph save/load

### PR DESCRIPTION
**Pull request by VoVAllen**
## Description

Breaking Change:
- `List` will automatically call `Value.data` for `Value` object, to get the real data
- `save_graphs` for `DGLHeteroGraph` no longer supports save labels with graphs

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

**Comment by VoVAllen** [comment]
@jermainewang Ready for review

File: python/dgl/data/graph_serialize.py [hunk 1]
```diff
@@ -108,43 +118,56 @@ def save_graphs(filename, g_list, labels=None):
...
-    return [gdata.get_graph() for gdata in metadata.graph_data], label_dict
-
+    if metadata.is_hetero:
+        g_list = [gdata.get_graph() for gdata in metadata.hetero_graph_data]
+    else:
+        for k, v in metadata.labels.items():
+            label_dict[k] = F.zerocopy_from_dgl_ndarray(v)
+        g_list = [gdata.get_graph() for gdata in metadata.graph_data], label_dict
+    if ignore_labels or len(label_dict) == 0:
+        return g_list
```
Comment (classicsong) [inline python/dgl/data/graph_serialize.py:168]: Can we return g_list, None?

File: tests/compute/test_serialize.py [hunk 2]
```diff
@@ -174,13 +177,103 @@ def test_serialize_empty_dict():
...
+    load_edges = load_g.all_edges('uv', 'eid')
+    assert np.allclose(F.asnumpy(load_edges[0]), edges0)
+    assert np.allclose(F.asnumpy(load_edges[1]), edges1)
+
+
+def create_heterographs(index_dtype):
+    g_x = dgl.graph(([0, 1, 2], [1, 2, 3]), 'user',
+                    'follows', index_dtype=index_dtype, restrict_format='any')
+    g_y = dgl.graph(([0, 2], [2, 3]), 'user', 'knows', index_dtype=index_dtype, restrict_format='csr')
+    g_x.nodes['user'].data['h'] = F.randn((4, 3))
```
Comment (classicsong) [inline tests/compute/test_serialize.py:221]: What will happen if there is an empty data e.g., data['h'] = None?

File: src/graph/serialize/dglgraph_serialize.cc [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+  fs->Write(node_tensors);
+  fs->Write(edge_tensors);
+}
+
+bool GraphDataObject::Load(dmlc::Stream *fs) {
+  NDArray indptr, indices, edge_ids;
+  fs->Read(&indptr);
+  fs->Read(&indices);
+  fs->Read(&edge_ids);
+  this->gptr = ImmutableGraph::CreateFromCSR(indptr, indices, edge_ids, "in");
```
Comment (classicsong) [inline src/graph/serialize/dglgraph_serialize.cc:199]: If the graph have restricted format, shall we re-build that format during load?
Also save the graph format preference.

File: python/dgl/data/graph_serialize.py (hunk 1 above)
Comment (VoVAllen) [reply python/dgl/data/graph_serialize.py:168]: In the previous dgl, it returns g_list, None. However from now on, we decided not to store labels with the graphs for simplicity. `save_graph(g_list)` no longer accepts `label_dicts`.
Therefore it doesn't make too much sense to return another value (None) for the future user?

File: tests/compute/test_serialize.py (hunk 2 above)
Comment (VoVAllen) [reply tests/compute/test_serialize.py:221]: It's not viable to assign `None` to ndata of heterograph. It must be a backend tensor

File: src/graph/serialize/dglgraph_serialize.cc (hunk 3 above)
Comment (VoVAllen) [reply src/graph/serialize/dglgraph_serialize.cc:199]: This one is for the legacy storage for `DGLGraph`, which we decided to store Immutable graph for all `DGLGraph`. For the detailed implementation of `HeteroGraph` you can see from https://github.com/VoVAllen/dgl/blob/c2e61ce10b4704ce24133900731787c0bcc5847e/src/graph/unit_graph.cc#L1458 and https://github.com/VoVAllen/dgl/blob/c2e61ce10b4704ce24133900731787c0bcc5847e/src/graph/heterograph.cc#L396

Comment (VoVAllen) [reply src/graph/serialize/dglgraph_serialize.cc:199]: One question would be  whether the save and load implementation on `UnitGraph` and `HeteroGraph` would be a stable(won't change in the near feature) implementation? @jermainewang

**Comment by jermainewang** [comment]
Could you remove those binary files such as `tests/compute/data/1.bin`?

**Comment by VoVAllen** [comment]
@jermainewang They are pretty small and used to test backward compatibility. I think it's fine to put it there?

File: python/dgl/data/graph_serialize.py [hunk 4]
```diff
@@ -108,42 +119,55 @@ def save_graphs(filename, g_list, labels=None):
...
-    graph_list: list of immutable DGLGraphs
-    labels: dict of labels stored in file (empty dict returned if no
+
+    graph_list: list of DGLGraphs / DGLHeteroGraph
+    labels(Optional): dict of labels stored in file (empty dict returned if no
     label stored)
 
+    If the file is stored with labels (before DGL 0.4.4) and set ignore_labels=False,
+     it will return graph_list and labels
+    If the file isn't stored with labels, it will only return graph_list
```
Comment (jermainewang) [inline python/dgl/data/graph_serialize.py:145]: Why not just checking the version number and returns w/ or w/o labels correspondingly?

File: python/dgl/data/graph_serialize.py [hunk 5]
```diff
@@ -108,42 +119,55 @@ def save_graphs(filename, g_list, labels=None):
...
+
     Examples
     ----------
     Following the example in save_graphs.
 
     >>> from dgl.data.utils import load_graphs
-    >>> glist, label_dict = load_graphs("./data.bin") # glist will be [g1, g2]
-    >>> glist, label_dict = load_graphs("./data.bin", [0]) # glist will be [g1]
+    >>> glist = load_graphs("./data.bin") # glist will be [g1, g2]
+    >>> glist = load_graphs("./data.bin", [0]) # glist will be [g1]
```
Comment (jermainewang) [inline python/dgl/data/graph_serialize.py:154]: You should keep the old example and add a note like "Before DGL v0.4, ...; Starting with DGL v0.4, ...".

File: tests/compute/test_serialize.py [hunk 6]
```diff
@@ -4,6 +4,7 @@
 import time
 import tempfile
```
Comment (jermainewang) [inline tests/compute/test_serialize.py:5]: @aksnzhy This tempfile package looks like exactly what you need.

File: tests/compute/test_serialize.py [hunk 7]
```diff
@@ -174,13 +176,103 @@ def test_serialize_empty_dict():
...
+    assert np.allclose(F.asnumpy(labels0['label']), np_labels)
+    assert np.allclose(F.asnumpy(labels1['label']), np_labels)
+
+    load_g = loadg_list[idx]
+    load_edges = load_g.all_edges('uv', 'eid')
+    assert np.allclose(F.asnumpy(load_edges[0]), edges0)
+    assert np.allclose(F.asnumpy(load_edges[1]), edges1)
+
+
+def create_heterographs(index_dtype):
```
Comment (jermainewang) [inline tests/compute/test_serialize.py:216]: Checkout the `graph_cases` under `test_utils`. We should gradually consolidate our unittests to use graphs there.

File: tests/compute/test_serialize.py [hunk 8]
```diff
@@ -174,13 +176,103 @@ def test_serialize_empty_dict():
...
+    edges = g_list[0]['follows'].edges()
+    assert np.allclose(F.asnumpy(edges[0]), np.array([0, 1, 2]))
+    assert np.allclose(F.asnumpy(edges[1]), np.array([1, 2, 3]))
+
+
+def test_serialize_heterograph():
+    f = tempfile.NamedTemporaryFile(delete=False)
+    path = f.name
+    f.close()
+    g_list0 = create_heterographs("int64") + create_heterographs("int32")
```
Comment (jermainewang) [inline tests/compute/test_serialize.py:247]: We probably need more test cases such as empty graph, graph with isolated nodes, graph with empty relations, etc. You could add them to `graph_cases.py`.

**Comment by jermainewang** [comment]
> @jermainewang They are pretty small and used to test backward compatibility. I think it's fine to put it there?

I see. That's fine.

File: python/dgl/data/graph_serialize.py [hunk 9]
```diff
@@ -70,32 +73,40 @@ def save_graphs(filename, g_list, labels=None):
         File name to store DGLGraphs. 
     g_list: list
         DGLGraph or list of DGLGraph
-    labels: dict (Default: None)
-        labels should be dict of tensors/ndarray, with str as keys
 
     Examples
     ----------
+    .. warning:: From DGL 0.4.4, save_graphs no longer supports saving tensor dict(labels) with graphs. If
```
[the remark this answers is not in the document]
Comment (VoVAllen) [reply python/dgl/data/graph_serialize.py:79]: Yes

File: python/dgl/container.py [hunk 10]
```diff
@@ -64,10 +72,12 @@ class StrMap(Map):
 
     You can use convert to create a dict[str->ObjectBase] into a Map.
     """
+
     def items(self):
         """Get the items from the map"""
         akvs = _api_internal._MapItems(self)
-        return [(akvs[i].data, akvs[i+1]) for i in range(0, len(akvs), 2)]
+        return [(akvs[i], akvs[i+1]) for i in range(0, len(akvs), 2)]
```
[the remark this answers is not in the document]
Comment (VoVAllen) [reply python/dgl/container.py:79]: Not yet.

**Comment by jermainewang** [comment]
@VoVAllen what's the progress of this PR?

**Comment by VoVAllen** [comment]
@jermainewang Ready for review.

Some changes:

- Now supports storing labels with DGLHeteroGraph, as someone told me this is useful in dataset construction part
- Tested and Built with S3 support now
- Not tested/built with HDFS support yet since I don't know how to use it, but I believe it's easy to add it just by changing the compilation options

File: python/dgl/data/graph_serialize.py [hunk 11]
```diff
@@ -115,16 +124,18 @@ def load_graphs(filename, idx_list=None):
...
 
     Returns
     ----------
-    graph_list: list of immutable DGLGraphs
+
+    graph_list: list of DGLGraphs / DGLHeteroGraph
+
     labels: dict of labels stored in file (empty dict returned if no
-    label stored)
+    label stored). Always the full data regardless of the idx_list
```
Comment (jermainewang) [inline python/dgl/data/graph_serialize.py:138]: This will not correctly render. Change to the following
```python
Returns
--------
graph_list: list of DGLGraphs / DGLHeteroGraph
    The loaded graphs.
labels: dict[str, Tensor]
    The graph labels stored in file. If no label is stored, the dictionary is empty.
    Regardless of whether the ``idx_list`` argument is given or not, the returned dictionary
    always contains labels of all the graphs.
```

File: python/dgl/data/graph_serialize.py [hunk 12]
```diff
@@ -54,40 +57,34 @@ def get_graph(self):
...
     Parameters
     ----------
     filename : str
-        File name to store DGLGraphs. 
+        File name to store graphs. 
     g_list: list
-        DGLGraph or list of DGLGraph
-    labels: dict (Default: None)
-        labels should be dict of tensors/ndarray, with str as keys
+        DGLGraph or list of DGLGraph/DGLHeteroGraph
```
Comment (jermainewang) [inline python/dgl/data/graph_serialize.py:75]: Where is the `labels` argument?

File: python/dgl/data/graph_serialize.py [hunk 13]
```diff
@@ -54,40 +57,34 @@ def get_graph(self):
...
-    >>> g1.add_nodes(3)
-    >>> g1.add_edges([0, 0, 0, 1, 1, 2], [0, 1, 2, 1, 2, 2])
-    >>> g1.ndata["e"] = th.ones(3, 5)
-    >>> g2 = dgl.DGLGraph()
-    >>> g2.add_nodes(3)
-    >>> g2.add_edges([0, 1, 2], [1, 2, 1])
-    >>> g2.edata["e"] = th.ones(3, 4)
+    >>> g1 = dgl.graph(([0, 1, 2], [1, 2, 3])
+    >>> g2 = dgl.graph(([0, 2], [2, 3])
+    >>> g2.edata["e"] = th.ones(2, 4)
```
Comment (jermainewang) [inline python/dgl/data/graph_serialize.py:87]: Add an example with labels argument.

File: python/dgl/data/heterograph_serialize.py [hunk 14]
```diff
@@ -0,0 +1,57 @@
+"""For HeteroGraph Serialization"""
```
Comment (jermainewang) [inline python/dgl/data/heterograph_serialize.py:1]: Please add some brief descriptions for each function and class in this file.

File: src/graph/serialize/graph_serialize.h [hunk 15]
```diff
@@ -6,71 +6,59 @@
...
- public:
-  DGL_DEFINE_OBJECT_REF_METHODS(GraphData, runtime::ObjectRef, GraphDataObject);
-
-  /*! \brief create a new GraphData reference */
-  static GraphData Create() {
-    return GraphData(std::make_shared<GraphDataObject>());
-  }
+enum FileVersion : uint64_t{
+  kV0 = 0ull,
+  kV1 = 1ull
```
Comment (jermainewang) [inline src/graph/serialize/graph_serialize.h:44]: Is this used? Delete if not.

File: src/graph/serialize/graph_serialize.h [hunk 16]
```diff
@@ -6,71 +6,59 @@
...
 };
 
+constexpr uint64_t kDGLSerializeMagic = 0xDD2E4FF046B4A13F;
+
 class StorageMetaDataObject : public runtime::Object {
  public:
+  bool is_hetero;
```
Comment (jermainewang) [inline src/graph/serialize/graph_serialize.h:51]: Is this still needed?

File: src/graph/serialize/streamwithcount.h [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+ * \file graph/serialize/streamwithcount.h
+ * \brief Graph serialization header
+ */
+#ifndef DGL_GRAPH_SERIALIZE_STREAMWITHCOUNT_H_
+#define DGL_GRAPH_SERIALIZE_STREAMWITHCOUNT_H_
+
+#include <dmlc/io.h>
+#include <dmlc/type_traits.h>
+
+class StreamWithCount : dmlc::Stream {
```
Comment (jermainewang) [inline src/graph/serialize/streamwithcount.h:12]: Please add docstring and a simple example for usage.

File: src/graph/serialize/streamwithcount.h [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+                                 bool allow_null = false) {
+    return new StreamWithCount(uri, flag, allow_null);
+  }
+
+  size_t Read(void *ptr, size_t size) { return strm_->Read(ptr, size); }
+
+  void Write(const void *ptr, size_t size) {
+    count_ += size;
+    strm_->Write(ptr, size);
+  }
```
Comment (jermainewang) [inline src/graph/serialize/streamwithcount.h:24]: If they override the methods in base class, use the `override` qualifier.

File: src/graph/serialize/streamwithcount.h [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+
+  using dmlc::Stream::Read;
+  using dmlc::Stream::Write;
+
+  bool isValid() { return strm_.get(); }
```
Comment (jermainewang) [inline src/graph/serialize/streamwithcount.h:29]: What is this naming style?

File: src/graph/serialize/streamwithcount.h [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+
+  uint64_t count() { return count_; }
```
Comment (jermainewang) [inline src/graph/serialize/streamwithcount.h:31]: `uint64_t count() const`

File: src/graph/serialize/streamwithcount.h [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+ */
+#ifndef DGL_GRAPH_SERIALIZE_STREAMWITHCOUNT_H_
+#define DGL_GRAPH_SERIALIZE_STREAMWITHCOUNT_H_
+
+#include <dmlc/io.h>
+#include <dmlc/type_traits.h>
+
+class StreamWithCount : dmlc::Stream {
+ public:
+  static StreamWithCount *Create(const char *uri, const char *const flag,
```
Comment (jermainewang) [inline src/graph/serialize/streamwithcount.h:14]: docstring

File: src/graph/serialize/heterograph_serialize.cc [hunk 22]
```diff
@@ -0,0 +1,297 @@
...
+ * The storage structure is
+ * {
+ *   // MetaData Section
+ *   uint64_t kDGLSerializeMagic
+ *   uint64_t kVersion = 2
+ *   uint64_t GraphType = kDGLHeteroGraph
+ *   ** Reserved Area till 4kB **
+ *
+ *   dgl_id_t num_graphs
+ *   vector<dgl_id_t> graph_indices (start address of each graph)
```
Comment (jermainewang) [inline src/graph/serialize/heterograph_serialize.cc:16]: Is this still accurate? You probably need to add some descriptions about how to read index data stored at the end of the file.

File: src/graph/serialize/heterograph_serialize.cc [hunk 23]
```diff
@@ -0,0 +1,297 @@
...
+ *
+ *   vector<GraphData> graph_datas;
+ *
+ * }
```
Comment (jermainewang) [inline src/graph/serialize/heterograph_serialize.cc:19]: You might want to talk about label data too.

File: src/graph/serialize/utils.h [hunk 24]
```diff
@@ -0,0 +1,37 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file graph/serialize/utils.h
+ * \brief Graph serialization header
+ */
```
Comment (jermainewang) [inline src/graph/serialize/utils.h:5]: This header looks quite redundant. Consider merge it with the others.

**Comment by VoVAllen** [comment]
@jermainewang Done
